src/collapse/ghcprof.rs

```diff
@@ -0,0 +1,222 @@
+}
+
+/// `ghcprof` folder configuration options.
+#[derive(Clone, Debug, Default)]
+#[non_exhaustive]
+pub enum Source {
+    #[default]
+    /// The %time column
+    PercentTime,
+    /// The ticks column
+    Ticks,
+    /// The bytes column
+    Bytes,
+}
+
+/// A stack collapser for the output of `ghc`'s prof files.
+///
+/// To construct one, either use `ghcprof::Folder::default()` or create an [`Options`] and use
+/// `ghcprof::Folder::from(options)`.
+#[derive(Clone, Default)]
+pub struct Folder {
+    /// Cost for the current stack frame.
+    current_cost: usize,
+
+    /// Function on the stack in this entry thus far.
+    stack: Vec<String>,
+
+    opt: Options,
+}
+
+// the first character and last + 1
+#[derive(Debug)]
+struct Cols {
+    cost_centre: usize,
+    module: usize,
+    source: usize,
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: io::BufRead,
+        W: io::Write,
+    {
+        // Consume the header...
+        let mut line = Vec::new();
+        let cols = loop {
+            line.clear();
+            if reader.read_until(b'\n', &mut line)? == 0 {
+                warn!("File ended before start of call graph");
+                return Ok(());
+            };
+            let l = String::from_utf8_lossy(&line);
+
+            if l.split_whitespace()
+                .take(START_LINE.len())
+                .eq(START_LINE.iter().cloned())
+            {
+                let cost_centre = 0;
+                let module = l.find(START_LINE[2]).unwrap_or(0);
+                // Pick out these fixed columns, first two are individual only
+                // "%time %alloc   %time %alloc"
+                // `ticks` and `bytes` columns might appear on the end
+                // ticks header is right aligned
+                // bytes header tries to be right aligned but has a max width
+                // "%time %alloc   %time %alloc  ticks  bytes"
+                let source = match self.opt.source {
+                    Source::PercentTime => l.find("%time").unwrap(),
+                    // ticks and bytes columns are weirdly aligned so find the end of the col before
+                    Source::Ticks => l.rfind("%alloc").unwrap() + 6,
+                    Source::Bytes => l.rfind("ticks").unwrap() + 5,
+                };
+                break Cols {
+                    cost_centre,
+                    module,
+                    source,
+                };
+            }
+        };
+        // Skip one line
+        reader.read_until(b'\n', &mut line)?;
+
+        // Process the data...
+        let mut occurrences = Occurrences::new(1);
+        loop {
+            line.clear();
+            if reader.read_until(b'\n', &mut line)? == 0 {
+                break;
+            }
+            let l = String::from_utf8_lossy(&line);
+            let line = l.trim_end();
+            if line.is_empty() {
+                break;
```

Comment:
Is it impossible in this format for there to be empty lines in the data? As in, between rows of profiling information.

Reply:
added a comment to say it's not expected to contain blank lines